vm/vtpm-mitigation/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+        }
+        Err(_) => {
+            STATUS_RECOVERY_FAILED
+        }
+    }
+}
+
+/// Check if the given blob is a valid TPM blob
+fn is_valid_tpm_blob(blob: &[u8]) -> Result<(), usize> {
+    let result = check_blob(blob);
+    if let Err(size) = result {
+        Err(size)
+    } else {
+        Ok(())
+    }
+}
+
+/// Recover the given blob and return the recovered blob
+fn recover_vtpm_blob(original_blob: &[u8]) -> Result<Vec<u8>, bool> {
+    let _guard = RECOVER_MUTEX.lock();
+    if original_blob.len() > LEGACY_VTPM_SIZE {
+        tracing::error!("Blob size is greater than the legacy vtpm size, skipping recovery");
+        return Err(false);
+    }
+
+    if check_blob(original_blob).is_ok() {
+        tracing::info!("TPM NVRAM is already good, skipping recovery");
+        return Err(false);
+    }
+
+    let mut blob = Vec::from(original_blob);
+
+    tpm::recover::recover_blob(blob.as_mut());
+    let pending_nvram = Arc::new(Mutex::new(Vec::new()));
+    let plat = TpmPlatformCallbacks {
+        pending_nvram: pending_nvram.clone(),
+        time: Instant::now(),
+    };
+
+    let state_cow: Cow<'_, [u8]> = Cow::Borrowed(&blob);
+
+    let tpm_engine = MsTpm20RefPlatform::initialize(
+        Box::new(plat),
+        ms_tpm_20_ref::InitKind::ColdInitWithPersistentState {
+            nvmem_blob: state_cow,
+        },
+    );
+
+    if tpm_engine.is_err() {
+        tracing::error!("Failed to recover the blob, error: {:?}", tpm_engine);
+        return Err(false);
+    }
+
+    let mut tpm_engine_helper = TpmEngineHelper {
+        tpm_engine: tpm_engine.unwrap(),
+        reply_buffer: [0u8; 4096],
+    };
+
+    let result = tpm_engine_helper.initialize_tpm_engine();
+    if result.is_err() {
+        tracing::error!("Failed to initialize the tpm engine, error: {:?}", result);
+        return Err(false);
+    }
+
+    let result = tpm_engine_helper.allocate_guest_attestation_nv_indices(0, true, false, true);
+    if result.is_err() {
+        tracing::error!(
+            "Failed to allocate the guest attestation nv indices, error: {:?}",
+            result
+        );
+        return Err(false);
+    }
+
+    drop(tpm_engine_helper);
+
+    let data = {
+        let mut pending_nvram = pending_nvram.lock();
+        std::mem::take(&mut *pending_nvram)
+    };
+
+    if data.len() != blob.len() {
+        tracing::error!("Recovered blob size is not same as the original blob size");
+        return Err(false);
+    }
+    Ok(data)
+}
+
+
+
+
+#[cfg(test)]
+mod tests {
+    use tracing_subscriber::{EnvFilter};
+
+    pub fn setup_logging() {
+        tracing_subscriber::fmt()
+            .with_env_filter(EnvFilter::from_default_env().add_directive(tracing::Level::TRACE.into()))
+            .init();
+    }
+
+    use super::*;
+    use std::io::Write;
+    #[test]
+    fn test_corrupted_tpm_state_is_revovered() {
+        setup_logging();
+        let path: &'static str ="./test-data/corrupted_blob.bin";
```

Comment:
I don't think you want to do fs reads in tests like this, should we use `include_bytes!` instead? @smalis-msft thoughts?